### PR TITLE
[FIX] web_editor: fix italic and underline in FF

### PR DESCRIPTION
**Pull request by sebgeelen**
When selecting the text content of a link
the underline and italic commands were not working in Firefox.

We re-implement the Browser command to fix this.

task-2667950

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: odoo/addons/base/models/ir_translation.py [hunk 1]
```diff
@@ -426,8 +427,13 @@ def _sync_terms_translations(self, field, records):
                     translations_to_match.append(translation)
 
             for translation in translations_to_match:
-                matches = get_close_matches(translation.src, terms, 1, 0.9)
-                src = matches[0] if matches else None
+                src_text_content = lxml.html.fromstring(translation.src).text_content()
+                terms_text_content = {}
+                for term in terms:
+                    terms_text_content[lxml.html.fromstring(term).text_content()] = term
```
Comment (mart-e) [inline odoo/addons/base/models/ir_translation.py:433]: move this above to avoid recomputing it for each item of `translations_to_match`

**Review by mart-e — COMMENTED** [review]
I think this is a better behaviour but it means we could have a scenario where we mark text as bold in English and it keeps in regular in the translations as the text content hasn't changed.
ccing @rco-odoo to be sure this is something we want to do.

File: odoo/addons/base/models/ir_translation.py (hunk 1 above)
Comment (sebgeelen) [reply odoo/addons/base/models/ir_translation.py]: Good point.
Code amended.

**Comment by dmo-odoo** [comment]
@mart-e As far as I know, the behavior you are describing with bold is how it works in 14.0 and prior because it used to use `b`, `i` and `u` html tags which are small enough to pass through the cracks of the fuzzy matching. We broke this in 15.0 with the new editor because we always use `span` with a `style` now rather than the short tags, meaning that setting a word in bold in 15.0 invalidates the translation. This PR aims to fix that to revert to the 14.0 and prior behavior for bold, italic and underline. However, what will change with this PR is that previous cases were styling would have triggered a translation invalidation (e.g. adding a `span` with a `style`) would now keep the translation. At least it is consistent, which was not the case in 14.0 and prior, where the invalidation of the translation was dependent on the *length* of the added html tags. We need a functionnal decision here.

File: odoo/addons/base/models/ir_translation.py [hunk 2]
```diff
@@ -425,9 +426,15 @@ def _sync_terms_translations(self, field, records):
                 else:
                     translations_to_match.append(translation)
 
+            terms_text_content = {}
+            if len(translations_to_match) > 0:
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_translation.py:430]:
```suggestion
            if translations_to_match:
```

File: odoo/addons/base/models/ir_translation.py [hunk 3]
```diff
@@ -425,9 +426,15 @@ def _sync_terms_translations(self, field, records):
...
+                for term in terms:
+                    terms_text_content[lxml.html.fromstring(term).text_content()] = term
+
             for translation in translations_to_match:
-                matches = get_close_matches(translation.src, terms, 1, 0.9)
-                src = matches[0] if matches else None
+                src_text_content = lxml.html.fromstring(translation.src).text_content()
+                matches = get_close_matches(src_text_content, set(terms_text_content.keys()), 1, 0.9)
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_translation.py:436]: Not good, as `set` makes this nondeterministic.  You can actually pass the dict itself in order to match against its keys.
```suggestion
                matches = get_close_matches(src_text_content, terms_text_content, 1, 0.9)
```

**Review by rco-odoo — COMMENTED** [review]
I understand the rationale and implementation in the case of HTML fields, and it clearly makes sense in that context.

But the change in `_sync_terms_translations()` is actually not 100% correct, since the method was designed to be _independent from_ the kind of text to translate.  There is absolutely no guarantee that the terms can be parsed in HTML.  The terms may come from XML, HTML, or any other kind of format.  It could be RST, or any proprietary format.

File: odoo/addons/base/models/ir_translation.py (hunk 2 above)
Comment (sebgeelen) [reply odoo/addons/base/models/ir_translation.py]: done

File: odoo/addons/base/models/ir_translation.py (hunk 3 above)
Comment (sebgeelen) [reply odoo/addons/base/models/ir_translation.py]: done

**Comment by sebgeelen** [comment]
> I understand the rationale and implementation in the case of HTML fields, and it clearly makes sense in that context.
> 
> But the change in `_sync_terms_translations()` is actually not 100% correct, since the method was designed to be _independent from_ the kind of text to translate. There is absolutely no guarantee that the terms can be parsed in HTML. The terms may come from XML, HTML, or any other kind of format. It could be RST, or any proprietary format.

@rco-odoo Regarding your comment.

After looking into it I don't think it's an issue.
The `Translation.src` are only parsed in HTML in order to compare their textual content differences. We still store the raw src value in the DB.
The lxml html parser (`lxml.html.fromstring()`) seems to accept all kind of input string without error, and the `.text_content()` getter will only return the text, which is exactly what we want. So we are able to compare the textual changes of a translation without taking html formatting into account.

If the input src string is not HTML, it should therefore still behave like before. 
In my my tests, RST, plain text, XML and HTML works without issues.

What do you think ?

**Comment by rco-odoo** [comment]
> After looking into it I don't think it's an issue. The `Translation.src` are only parsed in HTML in order to compare their textual content differences. We still store the raw src value in the DB. The lxml html parser (`lxml.html.fromstring()`) seems to accept all kind of input string without error, and the `.text_content()` getter will only return the text, which is exactly what we want. So we are able to compare the textual changes of a translation without taking html formatting into account.

Your argument actually is: _it does not crash_.  But that does not mean that it is correct.

Consider a format like Markdown, for instance.  The various bits of text will contain formatting elements like `*stuff*`, `_stuff_`, and so on.  And the "textual content" as you implemented it also contains these formatting elements.  The point is to get rid of them for the sake of comparison.  That's why it is not correct.

**Comment by sebgeelen** [comment]
> Consider a format like Markdown, for instance. The various bits of text will contain formatting elements like `*stuff*`, `_stuff_`, and so on. And the "textual content" as you implemented it also contains these formatting elements.

It indeed remove formatting only for xml and html format. 
For other formats (eg. markdown) it work the same as now in production. 

So it's not perfect.  But it is better than what we have in production right now, an never worst.
In my opinion this is a win. 

If we want to make this change "perfect" we need a list of all the supported format, to be able to check all of those and only keep the textual content. I'm not sure it is want we want, neither am I the one who should take this decision. 

@dmo-odoo , what is your take o this ?

**Comment by rco-odoo** [comment]
> > Consider a format like Markdown, for instance. The various bits of text will contain formatting elements like `*stuff*`, `_stuff_`, and so on. And the "textual content" as you implemented it also contains these formatting elements.
> 
> It indeed remove formatting only for xml and html format. For other formats (eg. markdown) it work the same as now in production.
> 
> So it's not perfect. But it is better than what we have in production right now, an never worst. In my opinion this is a win.
> 
> If we want to make this change "perfect" we need a list of all the supported format, to be able to check all of those and only keep the textual content. I'm not sure it is what we want, neither am I the one who should take this decision.

You don't need a list of the supported formats.
You need an API that developers can implement to get their expected behavior.

I suggest to apply the "text content" thing only for fields using `translate=xml_translate` or `translate=html_translate`.
That way, the behavior really remains as before for other kinds of translations in third-party code.

**Comment by sebgeelen** [comment]
This PR as been retargeted to saas-15.1 here : #81033

**Pull request closed without merging**